communication/src/firmware_update.cpp

```diff
@@ -0,0 +1,743 @@
+    }
+    const auto t1 = millis();
+    CHECK(callbacks_->finish_firmware_update(flags.value()));
+    stats_.processingTime += millis() - t1;
+    e->type(d.type());
+    e->code(CoapCode::CHANGED);
+    e->id(0); // Will be set by the message channel
+    e->token(d.token(), d.tokenSize());
+    if (!cancelUpdate) {
+        // Send stats to the server
+        spark::JSONBufferWriter w(e->payloadData(), e->maxPayloadSize());
+        w.beginObject();
+        w.name("recv_chunks").value(stats_.receivedChunks);
+        w.name("dup_chunks").value(stats_.duplicateChunks);
+        w.name("ooo_chunks").value(stats_.outOfOrderChunks);
+        w.name("sent_acks").value(stats_.sentChunkAcks);
+        w.name("proc_time").value((unsigned)stats_.processingTime);
+        w.endObject();
+        if (w.dataSize() > 0 && w.dataSize() <= e->maxPayloadSize()) {
+            e->payloadSize(w.dataSize());
+        }
+        *respId = &finishRespId_;
+    } else {
+        updating_ = false;
+    }
+    return 0;
+}
+
+int FirmwareUpdate::handleChunkRequest(const CoapMessageDecoder& d, CoapMessageEncoder* e, int** respId,
+        bool validateOnly) {
+    const auto chunkTime = millis();
+    const char* data = nullptr;
+    size_t size = 0;
+    unsigned index = 0;
+    CHECK(decodeChunkRequest(d, &data, &size, &index));
+    if (validateOnly) {
+        return 0;
+    }
+    if (!updating_) {
+        LOG(WARN, "Received UpdateChunk request but no update is in progress");
+        return SYSTEM_ERROR_INVALID_STATE;
+    }
+    ++stats_.receivedChunks;
+    if (index == 0 || index > chunkCount_) { // Chunk indices are 1-based
+        ERROR_MESSAGE("Invalid chunk index: %u", index);
+        return SYSTEM_ERROR_PROTOCOL;
+    }
+    if ((index < chunkCount_ && size != chunkSize_) ||
+            (index == chunkCount_ && (index - 1) * chunkSize_ + size != transferSize_)) {
+        ERROR_MESSAGE("Invalid chunk size: %u", (unsigned)size);
+        return SYSTEM_ERROR_PROTOCOL;
+    }
+    bool isDupChunk = false;
+    if (index <= chunkIndex_) {
+        isDupChunk = true;
+    } else if (index > chunkIndex_ + windowSize_) {
+        LOG(WARN, "Chunk is out of receiver window");
+    } else {
+        // Index of the chunk relative to the left edge of the receiver window (0-based)
+        index -= chunkIndex_ + 1;
+        // Position of the chunk bit in the bitmap
+        const size_t wordIndex = index / 32;
+        const unsigned bitIndex = index % 32;
+        uint32_t w = chunks_[wordIndex];
+        if (w & (1 << bitIndex)) {
+            isDupChunk = true;
+        } else {
+            w |= (1 << bitIndex);
+            chunks_[wordIndex] = w;
+            const size_t offs = fileOffset_ + index * chunkSize_; // Chunk offset in the file
+            if (index == 0) {
+                // Shift the receiver window
+                unsigned bits = 0;
+                while ((bits = trailingOneBits(chunks_[0]))) {
```

Comment:
Nit: we can probably simplify this a bit.

Reply:
We discussed this out-of-band, and it definitely would be more effective to shift the window by 32-bit words rather than arbitrary number of bits. I'm not sure if the code implementing that approach would actually be more straightforward and simple though :shrug: